python/util.py

```diff
@@ -0,0 +1,207 @@
+from hashlib import sha256
+from sha256 import generate_hash, IV
+from consts import PRIME, R2, M
+
+import struct
+
+
+def sha_compress_leaves(a: bytes, b: bytes):
+    assert (
+        len(a) + len(b) == 64
+    ), "Cannot use sha_compress directly unless the length of bytes is a multiple of 64"
+    return generate_hash(bytearray(list(a) + list(b)), compress_only=True)
+
+
+def sha_hash(a: bytes) -> bytes:
+    return sha256(a).digest()
+
+
+def hash_raw_pod(raw: list[int]) -> bytes:
+    u8s = u32_to_u8(raw)
+    chunk_size = 64  # 64 byte chunks
+    state: list[int] = IV
+    for idx in range(int(len(u8s) / chunk_size)):
+        block = u8s[idx * chunk_size : (idx + 1) * chunk_size]
+        state = u8_to_u32(
+            list(
+                generate_hash(bytearray(block), initial_state=state, compress_only=True)
+            )
+        )
+        state = swap_endian(state)
+
+    remainder = u8s[-(len(u8s) % chunk_size) :]
+    if len(remainder) > 0:
+        block = [0] * 64
+        block[: len(remainder)] = remainder[:]
+        state = u8_to_u32(
+            list(
+                generate_hash(bytearray(block), initial_state=state, compress_only=True)
+            )
+        )
+
+    return bytes(u32_to_u8(state))
+
+
+def swap_endian(l: list[int]) -> list[int]:
+    return [swap32(x) for x in l]
+
+
+def swap32(i: int):
+    return struct.unpack("<I", struct.pack(">I", i))[0]
+
+
+def u8_to_u32(u8s: list[int]) -> list[int]:
+    elems = 4
+    u32s = []
+    for idx in range(int(len(u8s) / elems)):
+        u32s.append(
+            int.from_bytes(bytes(u8s[idx * elems : (idx + 1) * elems]), "little")
+        )
+    return u32s
+
+
+def u32_to_u8(u32s: list[int]) -> list[int]:
+    u8s = []
+    for idx in range(len(u32s)):
+        u8s.extend(list(u32s[idx].to_bytes(4, "little")))
+    return u8s
+
+
+def to_elem(raw: int):
+    return encode_mont(raw % PRIME)
+
+
+def encode_mont(a: int) -> int:
+    return mul(R2, a)
+
+
+def decode_mont(a: int) -> int:
+    return mul(1, a)
+
+
+def add(lhs: int, rhs: int) -> int:
+    x = wrapped_add(lhs, rhs, 32)
+    return x if x < PRIME else wrapped_sub(x, PRIME, 32)
+
+
+def sub(lhs: int, rhs: int) -> int:
+    x = wrapped_sub(lhs, rhs, 32)
+    return x if x >= PRIME else wrapped_add(x, PRIME, 32)
```

Comment:
this line looks off, maybe it should be 
```suggestion
    return x if x < PRIME else wrapped_add(x, PRIME, 32)
```

Reply:
The suggestion looks right, I've no idea how I happened to get the right answers 🤔 will test locally